clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -91,8 +97,14 @@ void CodeGenTypes::addRecordTypeName(const RecordDecl *RD,
               OS << LlvmTyName;
             } else
               TTy->print(OS, false, true);
-          } else if (TemplateArg.getKind() == TemplateArgument::Integral)
-            OS << TemplateArg.getAsIntegral();
+          } else if (TemplateArgs[I].getKind() == TemplateArgument::Integral) {
+            const auto IntTemplateParam = TemplateArgs[I].getAsIntegral();
+            // Last parameter 'Use' is optional in SPIR-V, but is not optional
+            // in DPCPP headers. If it has Unnecessary value - skip it
+            constexpr size_t Unnecessary = 3;
```

Comment:
For my understanding can you explain this? What is the significance of '3' here?

Reply:
>>What is the significance of '3' here?

comment would be helpful. Thank you

Reply:
Added the comment

Reply:
I'm still a little confused. My question was  why is "3" = Unecessary. Is this defined by the spec? What are possible legal values for Use parameter?

Reply:
No, there is no `Unnecessary` in both SPIR-V and SYCL spec. Legal values on user-visible API level are 0, 1 and 2 (same for SPIR-V). Yet `Unnecessary` is defined in intermediate level of SYCL headers to keep compatibility with previous version of the extension, where we didn't have `Use` parameter making to be a default `Use` value.

While answering your question another idea came up in my mind, which might be cleaner. What if we declare `struct __spirv_JointMatrixINTEL` depending on `SYCL_EXT_ONEAPI_MATRIX` , WDYT @yubingex007-a11y @dkhaldi ? Like this:
```
#if SYCL_EXT_ONEAPI_MATRIX == 1
template <typename T, std::size_t R, std::size_t C, MatrixLayout L,
          Scope::Flag S = Scope::Flag::Subgroup>
struct __spirv_JointMatrixINTEL {
  T(*Value)
  [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1];
};
#else
template <typename T, std::size_t R, std::size_t C, MatrixLayout L,
          Scope::Flag S = Scope::Flag::Subgroup,
          MatrixUse U = MatrixUse::Unnecessary>
struct __spirv_JointMatrixINTEL {
  T(*Value)
  [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1]
     [static_cast<size_t>(U) + 1];
};
#endif // SYCL_EXT_ONEAPI_MATRIX
```

Reply:
I like this idea as long this does not break current SPIRV (use is still optional).

Reply:
Eh, there is quite a lot of things to change in the headers with a lot of code duplication to make it working. @elizabethandrews are we good to have the patch in clang or additional code commentary is required?

Reply:
Can we do something like:
```
template <typename T, std::size_t R, std::size_t C, MatrixLayout L,
          Scope::Flag S = Scope::Flag::Subgroup
#if SYCL_EXT_ONEAPI_MATRIX == 1
          ,MatrixUse U = MatrixUse::Unnecessary>
#else
>
#endif
struct __spirv_JointMatrixINTEL {
  T(*Value)
  [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1]
#if SYCL_EXT_ONEAPI_MATRIX == 1
     [static_cast<size_t>(U) + 1];
#else
  ;
#endif // SYCL_EXT_ONEAPI_MATRIX
};
```

Also, I see that Unnecessary template parameter is a default value, why can't we just declare it? How it will be incompatible with the previous version of the extension?

Also, can we do smth like:
```
template <typename T, std::size_t R, std::size_t C, MatrixLayout L,
          Scope::Flag S = Scope::Flag::Subgroup,
          MatrixUse U = MatrixUse::Unnecessary>
struct __spirv_JointMatrixINTEL {
  T(*Value)
  if constexpr (U == MatrixUse::Unnecessary) // or the same with a normal if?
    [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1];
  else
      [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1]
       [static_cast<size_t>(U) + 1];
};
```
?

Reply:
Can we do something like:
```
template <typename T, std::size_t R, std::size_t C, MatrixLayout L,
          Scope::Flag S = Scope::Flag::Subgroup
#if SYCL_EXT_ONEAPI_MATRIX == 1
          ,MatrixUse U = MatrixUse::Unnecessary>
#else
>
#endif
struct __spirv_JointMatrixINTEL {
  T(*Value)
  [R][C][static_cast<size_t>(L) + 1][static_cast<size_t>(S) + 1]
#if SYCL_EXT_ONEAPI_MATRIX == 1
     [static_cast<size_t>(U) + 1];
#else
  ;
#endif // SYCL_EXT_ONEAPI_MATRIX
};
```

Also, I see that Unnecessary template parameter is a default value, why can't we just declare it? How it will be incompatible with the previous version of the extension?